### PR TITLE
be more stingy

### PR DESCRIPTION
**Pull request by ymatsuda**
@andrewconner @ahsu1230 @stkem @eishay 
- added a different flavor of getCollectionsForKeep to KeepToCollectionRepo. It takes Keep instead of Id[Keep], so we don't have to access the bookmark table when we already have a Keep in memory. As the result, we do 2-way join instead of 3-way join.
- I noticed BasicCollectionCache is not invalidated when collections change. Invalidation is added to CollectionRepo and KeepToCollectionRepo.

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToCollectionRepo.scala [hunk 1]
```diff
@@ -68,6 +72,21 @@ class KeepToCollectionRepoImpl @Inject() (
...
 
+  def getCollectionsForKeep(keep: Keep)(implicit session: RSession): Seq[Id[Collection]] = {
+    if (keep.isActive) {
+      collectionsForKeepCache.getOrElse(CollectionsForKeepKey(keep.id.get)) {
+        val query = for {
+          kc <- rows if kc.bookmarkId === keep.id.get && kc.state === KeepToCollectionStates.ACTIVE
+          c <- collectionRepo.rows if c.id === kc.collectionId && c.state === CollectionStates.ACTIVE
+        } yield kc
+
+        query.sortBy(_.updatedAt).map(_.collectionId).list // todo(martin): we should add a column for explicit ordering of tags
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToCollectionRepo.scala:83]: I don't think Martin will ever get to that :(

**Comment by andrewconner** [comment]
Thanks, nice!

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToCollectionRepo.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToCollectionRepo.scala:83]: :-P

**Comment by stkem** [comment]
:+1:

**Comment by eishay** [comment]
good find. its a strange merge to start with, i would expect that the caller would not call it if the keep is not active.

**Pull request merged**